prow/config/config.go

```diff
@@ -659,6 +662,57 @@ type Deck struct {
 	// accepts a key of: `org/repo`, `org` or `*` (wildcard) to define what GitHub org (or repo) a particular
 	// config applies to and a value of: `RerunAuthConfig` struct to define the users/groups authorized to rerun jobs.
 	RerunAuthConfigs RerunAuthConfigs `json:"rerun_auth_configs,omitempty"`
+	// RestrictStoragePaths restricts artifact requests to specific buckets.
+	// By default, buckets listed in the GCSConfiguration are automatically allowed.
+	// Additional locations can be allowed via `AdditionalAllowedBuckets` fields.
+	RestrictStoragePaths *bool `json:"restrict_storage_paths,omitempty"`
+	// AdditionalAllowedBuckets is a list of storage buckets to allow in artifact requests
+	// (in addition to those listed in the GCSConfiguration).
+	// Setting this field requires "RestrictStoragePaths" also be set to `true`.
+	AdditionalAllowedBuckets []string `json:"additional_allowed_buckets,omitempty"`
+}
+
+// Validate performs validation and sanitization on the Deck object.
+func (d *Deck) Validate() error {
+	if len(d.AdditionalAllowedBuckets) > 0 && !d.ShouldValidateStoragePaths() {
+		return fmt.Errorf("deck.restrict_storage_paths is not enabled despite deck.additional_allowed_buckets being configured: %v", d.AdditionalAllowedBuckets)
+	}
+
+	// TODO(@clarketm): Remove "rerun_auth_config" validation in July 2020
+	if d.RerunAuthConfig != nil {
+		logrus.Warning("rerun_auth_config will be deprecated in July 2020, and it will be replaced with rerun_auth_configs['*'].")
+
+		if d.RerunAuthConfigs != nil {
+			return errors.New("rerun_auth_config and rerun_auth_configs['*'] are mutually exclusive")
+		}
+
+		d.RerunAuthConfigs = RerunAuthConfigs{"*": *d.RerunAuthConfig}
+	}
+
+	// Note: The RerunAuthConfigs logic isn't deprecated, only the above RerunAuthConfig stuff is
+	if d.RerunAuthConfigs != nil {
+		for k, config := range d.RerunAuthConfigs {
+			if err := config.Validate(); err != nil {
+				return fmt.Errorf("rerun_auth_configs[%s]: %v", k, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// ShouldValidateStoragePaths returns whether or not the Deck's storage path should be validated.
+// Validation could be either disabled by default or explicitly turned off.
+func (d *Deck) ShouldValidateStoragePaths() bool {
+	if d.RestrictStoragePaths == nil {
+		return RestrictStoragePathsByDefault
+	}
+	return *d.RestrictStoragePaths
+}
+
+// ShouldWarnAboutInvalidStoragePaths returns true unless the Deck is explicitly configured to *not* validate storage paths.
+func (d *Deck) ShouldWarnAboutInvalidStoragePaths() bool {
```

Comment:
Is there a reason to put this into this func that has exactly one caller rather than doing that check in this exactly one caller? If we want to test the logging, shouldn't we instead just do that rather than testing the func that we think will cause the logging to happen?

Reply:
As far as "reasons"... just my personal experience/style, I'd say. My preference is usually something resembling "single level of abstraction per function". Because (for example), it's easier to write confidence-building unit-tests for tiny methods like this vs "did Xxxlogging framework get called with the right varargs". But (as I mention above), I'll defer to your Go-lang sensibilities. :)

Reply:
Removed extraneous function `ShouldWarnAboutInvalidStoragePaths`.